### PR TITLE
Fix okhttp trustManager refelection not supported in java9

### PR DESCRIPTION
**Pull request by yuxuanchiadm**
Fixed UnsupportedOperationException occurs in jdk9.
Addition information: https://bugs.eclipse.org/bugs/show_bug.cgi?id=517113

### CODE REVIEW

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java [hunk 1]
```diff
@@ -61,18 +71,137 @@ public boolean equals(Object obj) {
...
+        this.sslSocketFactory = sslContext.getSocketFactory();
+        this.trustManager = (X509TrustManager) DefaultManagersHolder.trustManagers[0];
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  // Copied from sun.security.ssl.SSLContextImpl
+  private static final class DefaultManagersHolder {
```
Comment (fbricon) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java]: sun.security.ssl.SSLContextImpl from OpenJDK is GPLv2. I don't believe we can include it with EPL code

Comment (yuxuanchiadm) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java]: Oh you are right. EPL is not compatible with GPL. I'll go try implements it myself later. Or any other possible way we can get SSLContex and TrustManager both?

**Comment by yuxuanchiadm** [comment]
Ok. Code refactored.

**Comment by fbricon** [comment]
Build fails with

```
[ERROR] Failures:
[ERROR]   AetherMockWebserverConnectorTest.testArtifactDownloadViaSsl:223->downloadArtifact:289 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to repo (https://localhost:58051/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetAuthSslTest>GetTest.testDownloadArtifact:45 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58207/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetAuthSslTest>GetTest.testDownloadArtifactWithWait:120 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58205/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetAuthSslTest>GetTest.testDownloadCorrupted:104
[ERROR]   GetProxyAuthSslTest>GetTest.testDownloadArtifact:45 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58357/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetProxyAuthSslTest>GetTest.testDownloadArtifactWithWait:120 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58353/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetProxyAuthSslTest>GetTest.testDownloadCorrupted:104
[ERROR]   GetProxySslTest>GetTest.testDownloadArtifact:45 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58292/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetProxySslTest>GetTest.testDownloadArtifactWithWait:120 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58288/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetProxySslTest>GetTest.testDownloadCorrupted:104
[ERROR]   GetSslTest>GetTest.testDownloadArtifact:45 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58187/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetSslTest>GetTest.testDownloadArtifactWithWait:120 org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58172/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   GetSslTest>GetTest.testDownloadCorrupted:104
[ERROR]   PutAuthSslTest>PutTest.testArtifactUpload:33 Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58637/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   PutAuthSslTest>PutTest.testMetadataUpload:49->AetherTestCase.assertExpectations:174 assertion failed for gid/aid/version/maven-metadata.xml.md5
[ERROR]   PutSslTest>PutTest.testArtifactUpload:33 Could not transfer artifact gid:aid:extension:classifier:version from/to async-test-repo (https://localhost:58626/repo): sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[ERROR]   PutSslTest>PutTest.testMetadataUpload:49->AetherTestCase.assertExpectations:174 assertion failed for gid/aid/version/maven-metadata.xml.md5
[INFO]
```

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java [hunk 2]
```diff
@@ -59,34 +68,113 @@ public boolean equals(Object obj) {
...
+      this.trustManager = trustManager;
+    } else {
+      // can not use use jdk-default ssl socket factory like previous because
+      // we can not get trust manager without reflection. So create new ssl
+      // socket factory using system properties to create fake jdk-default
+      // ssl socket factory.
+      try {
+        String trustManagerFactoryAlgorithm = TrustManagerFactory.getDefaultAlgorithm();
+        // Clearly specify TrustManagerFactory provider to SunJSSE.
+        TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(trustManagerFactoryAlgorithm, "SunJSSE");
```
Comment (fbricon) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java]: This apparently is not compatible with [IBM JDK](https://bugs.eclipse.org/bugs/show_bug.cgi?id=515859#c18)

Comment (fbricon) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java]: `TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())` seems to work for Oracle, OpenJDK and IBM JDKs. 

@yuxuanchiadm can you update the PR?

Comment (fbricon) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java]: @snjeza does this work for you?

**Comment by fbricon** [comment]
@yuxuanchiadm any chance you could fix the failing tests too please?

**Comment by snjeza** [comment]
@fbricon yes, it works for me when using Oracle JDK, IBM JDK or Open JDK.

**Comment by yuxuanchiadm** [comment]
@fbricon I think most of failing tests may caused by credentials problem. Try check your jvm default keystore contains credentials you want. If we want fix those issues fundamentally. We should refactored all testcases using specific keystore instead of jvm default keystore.

**Comment by fbricon** [comment]
The tests fail on both my macOS and Ubuntu machines. But pass on master.
Indeed a test keystore is now needed for those tests

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java [hunk 3]
```diff
@@ -59,34 +68,112 @@ public boolean equals(Object obj) {
...
+        sslContext.init(keyManagers, trustManagers, null);
+        this.sslSocketFactory = sslContext.getSocketFactory();
+        this.trustManager = (X509TrustManager) trustManagers[0];
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private boolean caseInsensitiveEequals(String s1, String s2) {
```
Comment (fbricon) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnectorFactory.java]: caseInsensitiveEquals

**Comment by ifedorenko** [comment]
Takari requires signed [CLA](http://takari.io/support/TakariCLA.pdf) for all pull requests we accept.

**Comment by Pr0methean** [comment]
This is blocking https://bugs.eclipse.org/bugs/show_bug.cgi?id=517113.

**Comment by wind57** [comment]
At the moment this blocking Eclipse using jdk-9 at all...

**Comment by jimshowalter** [comment]
We upgraded to the latest okhttp3 and now we are getting this same error. But we aren't running Java 9 anywhere.

**Comment by cstamas** [comment]
This repo is being archived

**Pull request closed without merging**